### PR TITLE
Port to Python 3.6 compatibility

### PR DESCRIPTION
**Pull request by moshez**
The usual stuff (six, unicode carefulness, print()) plus:

* Move to incremental, versioneer had weird problems on 3.6
* im_self/__self__ for methods
* Move to saner way of testing functions, required code changes

### CODE REVIEW

**Comment by moshez** [comment]
@markrwilliams -- this branch is ready now.

File: docs/conf.py [hunk 1]
```diff
@@ -16,5 +16,5 @@
 project = module.__name__
 copyright = ', '.join((module.metadata['copyright'], module.metadata['author']))
 author = module.metadata['author']
-version = module.__version__
-release = module.__version__
+version = str(module.__version__)
```
Comment (markrwilliams) [inline docs/conf.py]: This looks like it should have gone into the incremental PR.  Can you back this out and do a new PR for it?

File: ncolony/client/tests/test_heart.py [hunk 2]
```diff
@@ -41,6 +41,20 @@ def _cleanup():
     case.addCleanup(_cleanup)
     os.environ = myEnv
 
+def _getSelf(method):
+    if hasattr(method, 'im_self'):
```
Comment (markrwilliams) [inline ncolony/client/tests/test_heart.py]: [`hasattr` is broken on Python 2](https://hynek.me/articles/hasattr/).  Use `getattr `instead:

```python
_missing = missing
self = getattr(method, 'im_self', _missing)
```

File: ncolony/client/tests/test_heart.py [hunk 3]
```diff
@@ -41,6 +41,20 @@ def _cleanup():
...
+        return method.im_self
+    if hasattr(method, '__self__'):
+        return method.__self__
+    raise TypeError("no self", method)
+
+def _getFunc(method):
+    if hasattr(method, 'im_func'):
```
Comment (markrwilliams) [inline ncolony/client/tests/test_heart.py]: As above.

File: ncolony/reaperlib.py [hunk 4]
```diff
@@ -1,79 +0,0 @@
-# Copyright (c) Moshe Zadka
```
Comment (markrwilliams) [inline ncolony/reaperlib.py]: Why did you remove this file?  Was it unused?

File: ncolony/ctllib.py [hunk 5]
```diff
@@ -77,7 +77,7 @@ def add(places, name, cmd, args, env=None, uid=None, gid=None, extras=None):
         details['gid'] = gid
     if extras is not None:
         details.update(extras)
-    content = json.dumps(details)
+    content = json.dumps(details).encode('utf-8')
```
Comment (markrwilliams) [inline ncolony/ctllib.py]: This happens throughout the code.  I suggest wrapping encoding and decoding JSON to and from bytes in a pair of functions.

File: ncolony/tests/functional_test.py [hunk 6]
```diff
@@ -92,7 +93,7 @@ def main(argv):
     _analyzeLogFile(LOG_FILE)
 
 def _analyzeLogFile(log_file):
-    lines = list(file(log_file))
+    lines = list(open(log_file))
```
Comment (markrwilliams) [inline ncolony/tests/functional_test.py]: This will generate a `ResourceWarning`.  A `tox` environment that runs with all warnings is a good thing to have, so you should wrap this and all other unclosed files with `with`.

File: ncolony/directory_monitor.py [hunk 7]
```diff
@@ -24,7 +25,7 @@ def checker(location, receiver):
     path = filepath.FilePath(location)
     files = set()
     filesContents = {}
-    def _check():
+    def _check(path):
```
Comment (markrwilliams) [inline ncolony/directory_monitor.py]: Why is this better than a closure?

File: ncolony/tests/helper.py [hunk 8]
```diff
@@ -0,0 +1,8 @@
+# Copyright (c) Moshe Zadka
+# See LICENSE for details.
+"""Helper functions"""
+import json
+
+def dumps2utf8(something):
```
Comment (markrwilliams) [inline ncolony/tests/helper.py]: If this is worth abstracting into a function for tests, it's worth using elsewhere.  Also, you should write a test for this!

File: ncolony/directory_monitor.py (hunk 7 above)
Comment (moshez) [reply ncolony/directory_monitor.py]: It is more amenable to testing. I can check that the path argument is properly passed in. This is probably not a great test, so I opened #49

File: ncolony/tests/helper.py (hunk 8 above)
Comment (moshez) [reply ncolony/tests/helper.py]: #50
